### PR TITLE
unix: don't block for io if any io handle is primed

### PR DESCRIPTION
**Pull request by saghul**
R=@bnoordhuis

### CODE REVIEW

**Comment by saghul** [comment]
`uv_pipe_connect` will defer any error to the callback, by priming the io handle with `uv__io_feed`. If called from a prepare callback (the last thing before blocking for i/o) the loop would block for ever.

**Comment by saghul** [comment]
Damn, when running the entire test suite (not this individual test) things got fucked up:

~~~~
#0  0x0000000000a77800 in pipe ()
(gdb) bt
#0  0x0000000000a77800 in pipe ()
#1  0x000000000044d2c4 in process_wait (vec=0x7ffdc1203b60, n=1, timeout=5000) at ../test/runner-unix.c:196
#2  0x00000000004065bd in run_test (test=0x241dc4e "pipe_connect_on_prepare", benchmark_output=0, 
    test_count=1) at ../test/runner.c:284
#3  0x0000000000405c9b in maybe_run_test (argc=2, argv=0x241dc20) at ../test/run-tests.c:175
#4  0x00000000004056ab in main (argc=2, argv=0x241dc20) at ../test/run-tests.c:56
(gdb) 
~~~~

**Comment by saghul** [comment]
I'm completely baffled here. I have no clue why this little change could make that blow up :-(

File: test/test-pipe-connect-prepare.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (bnoordhuis) [inline test/test-pipe-connect-prepare.c]: I would change the copyright.

File: test/test-pipe-connect-prepare.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+# define BAD_PIPENAME "/path/to/unix/socket/that/really/should/not/be/there"
+#endif
+
+
+static int close_cb_called = 0;
+static int connect_cb_called = 0;
+
+uv_pipe_t pipe;
+uv_prepare_t prepare;
+uv_connect_t conn;
```
Comment (bnoordhuis) [inline test/test-pipe-connect-prepare.c]: These need to be static to avoid symbol clashes.

Comment (bnoordhuis) [inline test/test-pipe-connect-prepare.c]: Oh, and maybe rename `pipe` to `pipe_handle`.  It doesn't conflict but it's also the name of a unistd.h function.

File: test/test-pipe-connect-prepare.c [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  int r;
+
+  r = uv_pipe_init(uv_default_loop(), &pipe, 0);
+  ASSERT(r == 0);
+
+  r = uv_prepare_init(uv_default_loop(), &prepare);
+  ASSERT(r == 0);
+  uv_prepare_start(&prepare, prepare_cb);
+
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
```
Comment (bnoordhuis) [inline test/test-pipe-connect-prepare.c]: Suggestion: check the return values of uv_prepare_start and uv_run.

File: test/test-pipe-connect-prepare.c (hunk 2 above)
Comment (saghul) [inline test/test-pipe-connect-prepare.c]: FML, thanks for putting me on the right track Ben!

**Comment by saghul** [comment]
@bnoordhuis fixed all the things, hopefully :-) PTAL.

**Comment by saghul** [comment]
Pingje :-)

File: test/test-pipe-connect-prepare.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+# define BAD_PIPENAME "/path/to/unix/socket/that/really/should/not/be/there"
+#endif
+
+
+static int close_cb_called = 0;
+static int connect_cb_called = 0;
+
+static uv_pipe_t pipe_handle;
+static uv_prepare_t prepare;
+static uv_connect_t conn;
```
Comment (bnoordhuis) [inline test/test-pipe-connect-prepare.c]: Teensy tiny suggestion: maybe rename `prepare` to `prepare_handle` and `conn` to `conn_req` for consistency.

**Comment by bnoordhuis** [comment]
LGTM

**Comment by saghul** [comment]
Thanks Ben! Addressed the suggestions and landed.

**Pull request merged**